xarray/core/variable.py

```diff
@@ -893,16 +893,27 @@ def reduce(self, func, dim=None, axis=None, keep_attrs=False,
         if dim is not None and axis is not None:
             raise ValueError("cannot supply both 'axis' and 'dim' arguments")
 
+        if 'cum' in func.__name__:
+            if dim is None:
+                raise ValueError("must supply a 'dim' argument to %s"%(func.__name__))
+
         if dim is not None:
             axis = self.get_axis_num(dim)
         data = func(self.data if allow_lazy else self.values,
                     axis=axis, **kwargs)
 
         removed_axes = (range(self.ndim) if axis is None
                         else np.atleast_1d(axis) % self.ndim)
-        dims = [dim for n, dim in enumerate(self.dims)
+        dims = [adim for n, adim in enumerate(self.dims)
                 if n not in removed_axes]
 
+        if 'cum' in func.__name__:
```

Comment:
Actually, I guess it's OK to keep this -- it is all we need to actually implement `cumsum`, after all, but let's do the shape check unilaterally, without looking at the function name.